test/e2e/simple_test.go

```diff
@@ -120,3 +120,63 @@ func TestFailingBuild(t *testing.T) {
 		t.Fatalf("watchBuild did not return expected error: %v", err)
 	}
 }
+
+func TestBuildLowTimeout(t *testing.T) {
+	clients := setup(t)
+
+	buildName := "build-low-timeout"
+	if _, err := clients.buildClient.builds.Create(&v1alpha1.Build{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: buildTestNamespace,
+			Name:      buildName,
+		},
+		Spec: v1alpha1.BuildSpec{
+			Timeout: "20s",
+			Steps: []corev1.Container{{
+				Name:    "lowtimeoutstep",
+				Image:   "ubuntu",
+				Command: []string{"/bin/bash"},
+				Args:    []string{"-c", "sleep 2000"},
+			}},
+		},
+	}); err != nil {
+		t.Fatalf("Error creating build: %v", err)
+	}
+
+	b, err := clients.buildClient.watchBuild(buildName)
+	if err == nil {
+		t.Fatalf("watchBuild did not return expected BuildTimeout error")
+	}
+
+	// verify reason for build failure is timeout
```

Comment:
Can we add a check that the build executed for 20s, +/- some wiggle room? That wiggle room will be pretty high today, since we only poll every 30s.

Reply:
Okay how about we check if build duration is between buildTimeout and buildTimeout + 40s(30s poll + 10s tolerance)?

Reply:
Yeah that sounds good. Add a TODO to reduce the wiggle room once we're more strict, and maybe run the test a few times to make sure it's not flaky.

Reply:
I ran into couple of edge cases while re-running the test. So I fixed couple of bugs in this PR.